### PR TITLE
Use the first paragraph, instead of cookie-cutter text, for rustdoc descriptions

### LINKED ISSUE #82283: rustdoc: improve web search snippets
## Boilerplate

Some (but not all) search result snippets for rust items include the boilerplate "1.0.0[−][src]". Example screenshots below are linked to the Google search results page they are taken from.

Search results for "rust slice" show the boilerplate:
[image](https://www.google.com/search?q=rust+slice)

Search results for "rust slice" omit the boilerplate:
[image](https://www.google.com/search?q=rust+error)

Per https://developers.google.com/search/docs/advanced/appearance/good-titles-snippets and https://developers.google.com/search/docs/advanced/crawling/special-tags#data-nosnippet, we can wrap this particular piece of boilerplate in `<span data-nosnippet>` to omit it from snippets.

## Low-information meta descriptions

Also, search engines commonly make snippets from the `<meta name="description" value="...">` tag. The tag rustdoc currently generates is very low-information. For instance:

> API documentation for the Rust `slice` mod in crate `std`

> API documentation for the Rust `Error` trait in crate `std`.

It would be much better to fill this tag with the first sentence (or two) of actual documentation. For instance:

> A dynamically-sized view into a contiguous sequence, [T]. Slices are a view into a block of memory represented as a pointer and a length.

> Error is a trait representing the basic expectations for error values, i.e., values of type E in Result<T, E>.

### PR DESCRIPTION
**Pull request by notriddle**
Partially addresses #82283.

### CODE REVIEW

**Comment by camelid** [comment]
Please add tests in `src/test/rustdoc/` for the changes to `description` output.

File: src/test/rustdoc/description.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#![crate_name = "foo"]
+//! # Description test crate
+//!
+//! This is the contents of the test crate docstring. It should not show up in the description.
+
+// @matches 'foo/index.html' '//meta[@name="description"]/@content' 'Description test crate'
```
Comment (camelid) [inline src/test/rustdoc/description.rs]: Since you're not using regex, `@has` is the way to go.

Comment (notriddle) [reply src/test/rustdoc/description.rs]: Okay, that's fixed.

File: src/test/rustdoc/description.rs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#![crate_name = "foo"]
+//! # Description test crate
+//!
+//! This is the contents of the test crate docstring.
+//! It should not show up in the description.
+
+// @has 'foo/index.html' '//meta[@name="description"]/@content' \
+//   'Description test crate'
```
Comment (jyn514) [inline src/test/rustdoc/description.rs:8]: The test doesn't match on the exact string, only that the output contains the test string somewhere.
```suggestion
// @has 'foo/index.html' '//meta[@name="description"]/@content' \
//   'Description test crate'
// @!has - '//meta[@name="description"]/@content' 'should not show up'
```

File: src/test/rustdoc/description.rs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+// @has 'foo/foo_mod/index.html' '//meta[@name="description"]/@content' \
+//   'First paragraph description.'
```
Comment (jyn514) [inline src/test/rustdoc/description.rs:11]:
```suggestion
// @has 'foo/foo_mod/index.html' '//meta[@name="description"]/@content' \
//   'First paragraph description.'
// @!has - '//meta[@name="description"]/@content' 'Second paragraph'
```

**Comment by jyn514** [comment]
> Fixes #82283

That's actually two different issues that should probably have been reported separately. This fixes "Low-information meta descriptions" but not "Boilerplate" (`1.0.0[−][src]`).

**Pull request merged**